spectral_cube/io/casa_wcs.py

```diff
@@ -14,24 +60,142 @@ def wcs_casa2astropy(ndim, coordsys):
+        header['OBSGEO-Y'] = obsgeo_alt * np.sin(obsgeo_lon) * np.cos(obsgeo_lat)
+        header['OBSGEO-Z'] = obsgeo_alt * np.sin(obsgeo_lat)
+
+    # World coordinates
+
+    # Find worldmap entries
+
+    worldmap = {}
+
+    for key, value in coordsys.items():
+        if key.startswith('worldmap'):
+            index = int(key[8:])
+            worldmap[index] = value
+
+    # Now iterate through the different coordinate types to populate the WCS
+
+    header['WCSAXES'] = np.max([np.max(idx) + 1 for idx in worldmap.values()])
+
+    # Initialize PC
+    for i in range(header['WCSAXES']):
+        for j in range(header['WCSAXES']):
+            header[f'PC{i+1}_{j+1}'] = 0.
+
+    for coord_type in ('direction', 'spectral', 'stokes', 'linear'):
+
+        for index in range(len(worldmap)):
+            if f'{coord_type}{index}' in coordsys:
+                break
+        else:
+            continue
+
+        data = coordsys[f'{coord_type}{index}']
+
+        if coord_type == 'direction':
+            idx1, idx2 = worldmap[index] + 1
+            header[f'CTYPE{idx1}'] = AXES_TO_CTYPE[data['system']][data['axes'][0]] + '-' + data['projection']
+            header[f'CTYPE{idx2}'] = AXES_TO_CTYPE[data['system']][data['axes'][1]] + '-' + data['projection']
+            header[f'CRPIX{idx1}'] = data['crpix'][0] + 1
+            header[f'CRPIX{idx2}'] = data['crpix'][1] + 1
+            header[f'CRVAL{idx1}'] = data['crval'][0]
+            header[f'CRVAL{idx2}'] = data['crval'][1]
+            header[f'CDELT{idx1}'] = data['cdelt'][0]
+            header[f'CDELT{idx2}'] = data['cdelt'][1]
+            header[f'CUNIT{idx1}'] = sanitize_unit(data['units'][0])
+            header[f'CUNIT{idx2}'] = sanitize_unit(data['units'][1])
+            header['LONPOLE'] = data['longpole']
+            header['LATPOLE'] = data['latpole']
+            if data.get('system') in EQUATORIAL_SYSTEMS:
+                header['RADESYS'] = RADESYS[data['conversionSystem']]
+                if data['conversionSystem'] in EQUINOX:
+                    header['EQUINOX'] = EQUINOX[data['conversionSystem']]
+            # NOTE: unclear if it is deliberate that the following is always
+            # ?_2 and ?_1 or whether it should depend on the index of the
+            # longitude.
+            if data.get('system') in EQUATORIAL_SYSTEMS:
+                header[f'PV{idx2}_1'] = 0.
+                header[f'PV{idx2}_2'] = 0.
+            header[f'PC{idx1}_{idx1}'] = data['pc'][0, 0]
+            header[f'PC{idx1}_{idx2}'] = data['pc'][0, 1]
+            header[f'PC{idx2}_{idx1}'] = data['pc'][1, 0]
+            header[f'PC{idx2}_{idx2}'] = data['pc'][1, 1]
+        elif coord_type == 'stokes':
+            idx = worldmap[index][0] + 1
+            header[f'CTYPE{idx}'] = 'STOKES'
+            header[f'CRVAL{idx}'] = data['crval'][0]
+            header[f'CRPIX{idx}'] = data['crpix'][0] + 1
```

Comment:
so... this means CASA stores CRPIX in zero-indexed, not one-indexed form?

Reply:
(this comment applies elsewhere too - I'm just observing, I don't think any code change or even commentary is needed)

Reply:
Yes it seems so, which would make sense internally for calculations.